Add datdot_milestone_2

ninabreznik opened the pull request:
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/LSRr7PCjBpEbKGh89) has been filled out for this milestone.
- [x] This pull request is being made by the same account as the accepted application (I think so, couldn't find the original anymore)
- [x] In case of acceptance, the payment will be transferred to the BTC/ETH payment address provided in the application. (I added a Polkadot address, but if payments are really only in BTC or ETH, I can update, please let me know)
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md).

Link to the application pull request: private

alxs commented:
Thanks for the delivery @ninabreznik. We'll look into it as soon as possible. Also, a reminder to include a link to your testing guide.

ninabreznik commented:
> Thanks for the delivery @ninabreznik. We'll look into it as soon as possible. Also, a reminder to include a link to your testing guide.

Thank you. I have recorded a short screencast and linked it in the comments. Just for each case sending the link also here https://watch.screencastify.com/v/ZSx4VaflQnJaM9S9xrIB

alxs commented:
Thank you Nina, that'll be useful in understanding the delivery. However, I rather meant the instructions to run your test suite. See https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md#testing-guide

alxs commented:
Thank you for the information. Please note that we generally don't accept deliveries without tests. Also, your next milestone for the grant is unrelated to the current deliverables.

In any case, I won't be the one evaluating your delivery - I'll leave it up to the person who does to continue the discussion.

ninabreznik commented:
@

> Thank you for the information. Please note that we generally don't accept deliveries without tests. Also, your next milestone for the grant is unrelated to the current deliverables.
> 
> In any case, I won't be the one evaluating your delivery - I'll leave it up to the person who does to continue the discussion.

Thanks. Maybe just a note to the person who will be evaluating the delivery. Also in our first milestone, we used the same principle for testing our code (using the scenarios) so we really didn't expect to change this for the second milestone, therefore no 'classic' tests here :)

cruikshankss commented:
@ninabreznik Hi Nina, my name is Ashley, I work at W3F, and I am evaluating your second milestone. 

Thank you for your delivery! I've studied your project a bit and I find it very interesting. I'm excited to learn more how you're building this bridge between dat and Substrate. I think it's cool you're building a project to incentivize distributed storage. 

I'm looking forward to running your simulation (in lieu of classic tests) to get a good understanding of your codebase.

ninabreznik commented:
> @ninabreznik Hi Nina, my name is Ashley, I work at W3F, and I am evaluating your second milestone. 
> 
> Thank you for your delivery! I've studied your project a bit and I find it very interesting. I'm excited to learn more how you're building this bridge between dat and Substrate. I think it's cool you're building a project to incentivize distributed storage. 
> 
> I'm looking forward to running your simulation (in lieu of classic tests) to get a good understanding of your codebase.

Nice to meet you, Ashley. Thank you for lovely words and please let us know if/when you need our input. 😊

cruikshankss commented:
Hi @ninabreznik, I had a look at your screencast. 

You didn't provide the commands, but I ran `git clone git@github.com:playproject-io/datdot-service.git`, then cd `datdot-service`, then `git checkout milestone_2` but that failed and I don't see the `milestone_2` branch on your GitHub either.

I would rather not assume the instructions, even the terminal commands, and wondering if you could please provide step-by-step instructions on how to test, per the grant agreement. You seems to skip some steps on your screencast and I'm not sure if I'm doing something wrong or if your testing guide is wrong. 

I am usually able to process evaluations much faster and more thoroughly when there are detailed testing guides that go extra slow, step-by-step. Thank you so much!

ninabreznik commented:
Hi, @cruikshankss sorry about that. I added a little comment to the milestone-2 document (see below), but I probably should have recorded another screencast to avoid the confusion. My bad.

`Here’s the screencast with a short demo of the code https://watch.screencastify.com/v/ZSx4VaflQnJaM9S9xrIB (in the meanwhile I merged the milestone_2 branch to main, so you can just clone the main branch`

I also updated the readme, so the step by step proces should be the one listed in the readme https://github.com/playproject-io/datdot-service#building

Sorry again, hope it works this time :)

cruikshankss commented:
@ninabreznik Thank you so much for the update! Sorry I missed that. Working on it ASAP!

cruikshankss commented:
I got the simulation to run! Thank you. Working on the evaluation next. Hope you have a nice day.

ninabreznik commented:
Very glad to hear. Please let me know if you will need any additional explanation. I know it is not easy to come to a completely new project and have to understand how it works and what exactly it does. We were trying hard to document how things work so far, it's only been the internal team working and having to understand the code for now. So, kudos to be able to do what you do. Not sure if it helps, but I made some diagrams explaining the flows and they are published [here](https://datdot.org/#doc)

cruikshankss commented:
Hi @ninabreznik! 

I have been really enjoying testing and learning about your project. I was new too SRML but now I think it's really cool and I super appreciate reading your chain logic as well since my javascript is better than my Rust. 

For deliverable 4, you write:

>Implementation of the chain logic (simulated as javascript, but all the extrinsics and params will stay the same, they will just need to be reimplemented in rust/substrate)

But for Milestone 3, I do not see any deliverable for implementing a node in rust/substrate? Are you still planning to write your own node/chain with this chain logic? Or are you rather building a browser bridge with Service Worker, server/API, and a defined standard?

If you build a browser bridge, what is the purpose and how does it relate to a Substrate node/chain? If you don't mind, please explain it like I'm five (ELI5) since I'm relatively new here. :)

cruikshankss commented:
Hello @ninabreznik, I have finished my evaluation and would just like the answer to my above question before accepting the milestone! Sorry for the delay and eager to finish this up for you! Cheers!

ninabreznik commented:
Hi, @cruikshankss, I am very glad to hear you enjoyed the testing our project.

So, let me answer your question regarding the rust code. We have agreed with the Web3 foundation to hand over only the JS version of the chain logic for milestone 2, We will of course continue with implementing also the rust/substrate, but this will not be part of W3F grant anymore, but will instead be part of the Polkadot Treasury funding we have received approx. 1 month ago.


The last milestone for W3F grant will be a browser bridge to give polkadotjs dapp developers a "hyper gateway" to access and use hypercore protocol for any data in the browser, so they can use the incentivized data hosting mechanism datdot provides.

Authoring hypercore based datasets require keypairs which we store in the users data vault.
The architeture of the data vault needs to be designed and prototyped to be compatible with and to give controlled access to
1. "hyper web dapps" via the browser bridge 
2. datdot service and datdot chain operations

Milestone 1 + 2 enabled us to clearly see all requirements for (1.)
The Browser Bridge Milestone is enabeling us to upgrade the data vault and refactor the project based on it.

This will allow us to have a solid foundation to implement the substrate based version of the datdot chain (but this will be part of the Polkadot Treasury fund we have just received)

Hope this helps, please let me know if we can try to explain it differently. We can go into more details or try from another perspective :)

cruikshankss commented:
@ninabreznik Thank you so much for the clarity in the previous comment, and congratulations, I have just accepted your M2! 

I looked for your invoice to forward to our invoices team, but I could not find it. Have you submitted your invoice, and if so, sorry, but would you mind resubmitting it via [our form](https://forms.gle/LSRr7PCjBpEbKGh89)? Thank you!

ninabreznik commented:
Yay, thanks a lot. I did sent the invoice. Maybe you can search under Vision Baker Ltd. If it is not to be found, I am happy to resubmit it :)

cruikshankss commented:
@ninabreznik Ah, yes, thanks! I see there were two invoices submitted for M2, three days apart, so I assume to forward the most recent one which includes the BIC and IBAN. Please let me know if this is wrong.

ninabreznik commented:
Yes, the latest it the correct one. Thanks so much. You made my day. 

If you are ever looking for a side gig, we will soon (June?) start looking for someone to help us implement the JS node in substrate, so let me know if you would by any chance be interested :)

cruikshankss commented:
Great, I forwarded your invoice to the invoices team! 

Thank you so much for the compliments and I'm glad to make your day! That does sound so interesting and fun, but I am exceedingly happy with my job here. Best of luck @ninabreznik and please let me know if you have any further questions!

ninabreznik commented:
Awesome. I am sure you are happy with the job and this is how it should be. Wasn't sure if it's full time, that is why I asked ;)

Keep up with good work and we talk again once we deliver the next milestone.

Thank you once again for taking time and reviewing our work!
